### PR TITLE
Skip messages without BrokerEntryMetadata

### PR DESCRIPTION
**Pull request by BewareMyPower**
### Motivation

When users upgrade from KoP (< 2.8.0) to KoP (>= 2.8.0), the topics produced by KoP (< 2.8.0) cannot be used, like https://github.com/streamnative/kop/issues/789. Therefore, the upgrade requires old version KoP users to delete these topics and create again if they want to reuse the same topic name.

This PR intends to skip these old messages that don't have a BrokerEntryMetadata part so that when users upgrade from KoP (< 2.8.0) to KoP (>= 2.8.0), the topics can be reused and the old messages will be skipped automatically.

### Modifications

- Add `MessageMetadataUtils#asyncFindPosition` that leverages `ManagedLedger#asyncFindPosition` to find the position associated with the offset. The most significant change is that `true` will be returned if the entry doesn't contain the BrokerEntryMetadata. Because it means the entry is older than the target entry in binary search.
- In `MessageMetadataUtils#getOffsetOfPosition`, return the 0 offset if the entry doesn't has the BrokerEntryMetadata.
- Add an `UpgradeTest`. First disable the `BrokerEntryMetadataInterceptor` and produce some messages, then restart the broker with `BrokerEntryMetadataInterceptor` enabled and produce some messages. There are 5 topics for test to cover different cases in binary search. And the tests cover the three cases of `LIST_OFFSETS` request (earliest, latest, timestamp)
  - `testOffsetsInSendCallback`: Verify the offsets in send callback are correct. When BrokerEntryMetadata is not enabled, the offset should be -1.
  - `testConsumeEarliest`: Run Kafka consumers to consume messages from earliest and verify the old messages are skipped.
  - `testConsumeLatest`: Run Kafka consumers to consume messages from latest. This tests's priority is lower because a new message will be sent in this test and it will affect other tests.
  - `testConsumeEarliest`: Verity listing offsets by timestamp work with corrupted old messages.

### CODE REVIEW

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/KopProtocolHandlerTestBase.java [hunk 1]
```diff
@@ -329,24 +335,28 @@ protected final void internalCleanup() throws Exception {
...
-            if (mockZooKeeper != null) {
-                mockZooKeeper.shutdown();
-            }
             if (sameThreadOrderedSafeExecutor != null) {
                 sameThreadOrderedSafeExecutor.shutdown();
             }
-            if (bkExecutor != null) {
-                bkExecutor.shutdown();
-            }
+            cleanupStorage();
```
Comment (Demogorgon314) [inline tests/src/test/java/io/streamnative/pulsar/handlers/kop/KopProtocolHandlerTestBase.java:341]: Why did we `cleanupStorage()` here ?  I think your original purpose is split broker and storage cleanups right?

Comment (BewareMyPower) [reply tests/src/test/java/io/streamnative/pulsar/handlers/kop/KopProtocolHandlerTestBase.java:341]: It's right. I'll remove this line and try to verify the old data still exists after restart.

**Comment by BewareMyPower** [comment]
TODO: Add an option to enable this feature

**Comment by BewareMyPower** [comment]
After confirming with @codelipenghui, I think we don't need an extra option to enable skipping messages. So this PR is ready to review, PTAL @Demogorgon314

**Review by Demogorgon314 — APPROVED** [review]
LGTM

**Pull request merged**